File: Jour_2/poo.py
class Employee:
    def __init__(self, rank, isFree=True):
        self.rank = rank
        self.isFree = isFree

    def handleCall(self):
        if self.isFree:
            self.isFree = False
            return True
        else:
            return False

    def getName(self):
        return type(self).__name__


class Respondent(Employee):
    def __init__(self, isFree=True):
        super().__init__(0, isFree)


class Manager(Employee):
    def __init__(self, isFree=True):
        super().__init__(1, isFree)


class Director(Employee):
    def __init__(self, isFree=True):
        super().__init__(2, isFree)


class CallCenter():
    def __init__(self):
        self.queue = ['call1', 'call2', 'call3']
        self.employees = [Respondent(), Respondent(), Manager(), Director()]

    def dispatchCall(self):
        while self.queue:
            call = self.queue.pop(0)
            handled = False

            for rank in range(3):
                for employee in self.employees:
                    if employee.rank == rank and employee.handleCall():
                        print(f"{call} handle by {employee.getName()}")
                        handled = True
                        break
                if handled:
                    break
            if not handled:
                print("all busy")

        print("empty queue")
        return


center = CallCenter()
center.dispatchCall()

File: Jour_2/POO.js
// Imagine you have a call center with three levels of employees: 
// respondent, manager, and director.
// An incoming telephone call must be first allocated to a respondent who is free.
// If the respondent can't handle the call, he or she must escalate the call to a manager.
// If the manager is not free or not able to handle it, then the call should be escalated to a director.
// Design the classes and data structures for this problem.
// Implement a method dispatchCall() which assigns a call to the first available employee.



class Respondent {
    job;
    occupy;

    constructor(job) {
        this.job = job;
        this.occupy = occupy;
    }

    dispatchCall(task) {
        if (!this.occupy) {
            console.log(`${this.job} est maintenant occupé.`);
            this.occupy = true;
            task.status = 'en cours';
            return task;
        } else {
            console.log(`${this.job} est déjà occupé. La tâche passe au manager...`);
            return new Manager(this.job).assignTask(task);
        }
    }
}

class Manager extends Respondent {

    constructor(job) {
        super(job)
    }

    dispatchCall() {
        if (!this.occupy) {
            console.log(`${this.job} est maintenant occupé.`);
            this.occupy = true;
            task.status = 'en cours';
            return task;
        } else {
            console.log(`${this.job} est déjà occupé. La tâche passe au directeur...`);
            return new Director(this.job).assignTask(task);
        }
    }
}

class Director extends Respondent {
  
    constructor(job) {
        super(job)
    }

    dispatchCall() {
        if (!this.occupy) {
            console.log(`${this.job} est maintenant occupé.`);
            this.occupy = true;
            task.status = 'en cours';
            return task;
        } else {
            console.log(`Tout le monde est déjà occupé. La tâche reste sans réponse`);
            return null;
        }
    }
}
